runtime/drivers/pinot/sqldriver/driver.go

```diff
@@ -0,0 +1,293 @@
+		return nil
+	}
+	switch pinotType {
+	case "INT":
+		// check if interface is string as it may be NaN
+		if reflect.TypeOf(r.results.Get(rowIdx, coldIdx)).String() == "string" {
+			return int32(math.NaN())
+		}
+		return r.results.GetInt(rowIdx, coldIdx)
+	case "LONG":
+		if reflect.TypeOf(r.results.Get(rowIdx, coldIdx)).String() == "string" {
+			return int64(math.NaN())
+		}
+		return r.results.GetLong(rowIdx, coldIdx)
+	case "FLOAT":
+		if reflect.TypeOf(r.results.Get(rowIdx, coldIdx)).String() == "string" {
+			return float32(math.NaN())
+		}
+		return r.results.GetFloat(rowIdx, coldIdx)
+	case "DOUBLE":
+		if reflect.TypeOf(r.results.Get(rowIdx, coldIdx)).String() == "string" {
+			return math.NaN()
+		}
+		return r.results.GetDouble(rowIdx, coldIdx)
+	case "STRING":
+		return r.results.GetString(rowIdx, coldIdx)
+	case "BYTES":
+		// return hex string as it is
+		return r.results.GetString(rowIdx, coldIdx)
+	case "BIG_DECIMAL":
+		return r.results.Get(rowIdx, coldIdx)
+	case "TIMESTAMP":
+		// convert iso8601 formatted string to time.Time
+		t, err := time.Parse("2006-01-02 15:04:05.0", r.results.GetString(rowIdx, coldIdx))
+		if err != nil {
+			return err
+		}
+		return t
+	case "BOOLEAN":
+		return r.results.Get(rowIdx, coldIdx).(bool)
+	default:
+		return reflect.TypeOf("")
+	}
+}
+
+// ParseDSN parses the DSN string to extract the controller address and basic auth credentials
+func ParseDSN(dsn string) (string, map[string]string, error) {
+	// validate dsn - it should be a valid URL, may contain basic auth credentials
+	u, err := url.Parse(dsn)
+	if err != nil {
+		return "", nil, fmt.Errorf("invalid DSN: %w", err)
+	}
+
+	var authHeader map[string]string
+	if u.User != nil {
+		uname := u.User.Username()
+		pwd, passwordSet := u.User.Password()
+		if uname == "" || !passwordSet {
+			return "", nil, fmt.Errorf("DSN should contain valid basic auth credentials")
+		}
+		// clear user info from URL so that u.String() doesn't include it
+		u.User = nil
+		authString := fmt.Sprintf("%s:%s", uname, pwd)
+		authHeader = map[string]string{
+			"Authorization": fmt.Sprintf("Basic %s", base64.StdEncoding.EncodeToString([]byte(authString))),
+		}
+	}
+	return u.String(), authHeader, nil
+}
+
+func completeQuery(query string, args []sqlDriver.NamedValue) (string, error) {
+	parts := strings.Split(query, "?")
+	if len(parts)-1 != len(args) {
+		return "", fmt.Errorf("mismatch in the number of placeholders and arguments")
+	}
+
+	var sb strings.Builder
+	for i, part := range parts {
+		sb.WriteString(part)
+		if i < len(args) {
+			argStr, err := formatArg(args[i].Value)
+			if err != nil {
+				return "", err
+			}
+			sb.WriteString(argStr)
+		}
+	}
+
+	return sb.String(), nil
+}
+
+func formatArg(value sqlDriver.Value) (string, error) {
+	switch v := value.(type) {
+	case string, *big.Int, *big.Float:
+		// For pinot types - STRING, BIG_INT and BIG_DECIMAL - enclose in single quotes
+		return fmt.Sprintf("'%v'", v), nil
```

Comment:
Shouldn't it escape single quotes in strings? (In standard SQL, it would entail replacing single quotes with two single quotes.) Otherwise, it's a very easy SQL injection vulnerability